### PR TITLE
Add error details

### PR DESCRIPTION
**Pull request by vkareh**
Custom Errors can now have an optional `details` property that allows
for arbitrary details to be added to an error object. This attribute is
a slice and allows for appending more details to the same error object.

### CODE REVIEW

**Comment by vkareh** [comment]
@zgalor @tzvatot - please take a look. Any suggestions welcomed!

File: errors.go [hunk 1]
```diff
@@ -151,6 +152,24 @@ func GetUserMessage(err error) string {
...
+	Details() []interface{}
+}
+
+// Details returns the error details
+func (c *customError) Details() []interface{} { return c.details }
+
+// GetDetails returns a slice of arbitratry details for all errors
+// If error is not `detailedError` returns nil
+func GetDetails(err error) []interface{} {
+	if detailedError, ok := err.(detailedError); ok {
```
Comment (zgalor) [inline errors.go]: maybe name the variable different than the interface. this is a bit confusing

File: errors.go [hunk 2]
```diff
@@ -151,6 +152,24 @@ func GetUserMessage(err error) string {
...
+		return detailedError.Details()
+	}
+
+	return nil
```
Comment (zgalor) [inline errors.go:170]: I am not sure if it's better to return `nil` or empty slice.
I think both can be tested for as `if len(details) == 0` so we can leave this as nil.
This discussion continues on line 253 with the usage of this function

@cben @tzvatot WDYT?

File: errors.go [hunk 3]
```diff
@@ -215,6 +235,41 @@ func (errorType ErrorType) UserErrorf(msg string, args ...interface{}) error {
 	}
 }
 
+// DetailsAdd adds a details element to an error
+func (errorType ErrorType) DetailsAdd(err error, details interface{}) error {
```
Comment (zgalor) [inline errors.go]: `AddDetails` sounds like better English

File: errors.go [hunk 4]
```diff
@@ -215,6 +235,41 @@ func (errorType ErrorType) UserErrorf(msg string, args ...interface{}) error {
...
+	c := new(customError)
+	c.error = errors.WithStack(err)
+	c.userMessage = GetUserMessage(err)
+
+	origDetails := GetDetails(err)
+	if origDetails != nil {
+		c.details = append(origDetails, details)
+	} else {
+		c.details = []interface{}{details}
+	}
```
Comment (zgalor) [inline errors.go]: ok. back to my original comment regarding `GetDetails`
If it would `return []interface{}{}` I think we could remove this `if` block and always append.
Am I right? or maybe this requires a `make([]interface{})` and in that case, it's not worth the change.

@cben @tzvatot

File: errors.go [hunk 5]
```diff
@@ -215,6 +235,41 @@ func (errorType ErrorType) UserErrorf(msg string, args ...interface{}) error {
...
+		c.errorType = errorType
+	} else {
+		c.errorType = GetType(err)
+	}
+
+	return c
+}
+
+// Details creates a new error with arbitrary details
+func (errorType ErrorType) Details(details interface{}) error {
```
Comment (zgalor) [inline errors.go]: I don`t see an actual usage for that.
Sadly, I think we would always have to use 2 lines if we want to add details.

```
err = weberr.Unauthorized.Wrapf(err, "user is unauthorized")
err = weberr.AddDetails(err, detailsObj)
```

Do you have a use case where this func is useful for?

File: errors_test.go [hunk 6]
```diff
@@ -118,3 +118,39 @@ func TestGetType(t *testing.T) {
...
+	tests := []struct {
+		err      error
+		expected []interface{}
+	}{
+		{nil, []interface{}{}},
+		{io.EOF, []interface{}{}},
+		{DetailsAdd(io.EOF, details), []interface{}{details}},
+		{NoType.Details(details), []interface{}{details}},
+		{DetailsAdd(BadRequest.Details(details), details), []interface{}{details, details}},
+		{DetailsAdd(DetailsAdd(DetailsAdd(io.EOF, details), details), details), []interface{}{details, details, details}},
```
Comment (zgalor) [inline errors_test.go]: try to add a use case with different types of details and see they don't override

**Review by zgalor — CHANGES_REQUESTED** [review]
Good work.
Some comments and suggestions

File: errors.go (hunk 1 above)
Comment (tzvatot) [reply errors.go]: IMO it's good this way

File: errors.go (hunk 2 above)
Comment (tzvatot) [reply errors.go:170]: I suggest returning an empty slice.

File: errors.go (hunk 1 above)
Comment (vkareh) [reply errors.go]: I renamed the interface and now it's a bit more consistent with the `userMessager` interface

File: errors.go (hunk 4 above)
Comment (vkareh) [reply errors.go]: This doesn't require a `make`, in fact, just returning `[]interface{}{}` in `GetDetails` removes this entire chunk into a one-liner

File: errors.go (hunk 3 above)
Comment (vkareh) [reply errors.go]: Done

File: errors.go (hunk 2 above)
Comment (vkareh) [reply errors.go:170]: either one will be captured by the json encoder `omitempty`, in which case it doesn't matter much. Returning an empty slice simplifies the code further down.

File: errors_test.go (hunk 6 above)
Comment (vkareh) [reply errors_test.go]: Done.

File: errors.go (hunk 5 above)
Comment (vkareh) [reply errors.go]: I don't have user-facing use case, no. But the `AddDetails` function relies on it as a safeguard if someone calls it with a `nil` error object (just like `UserWrapf` does).

In short, `Details` is to `AddDetails` what `UserErrorf` is to `UserWrapf`.

File: errors.go [hunk 7]
```diff
@@ -215,6 +235,36 @@ func (errorType ErrorType) UserErrorf(msg string, args ...interface{}) error {
...
+func (errorType ErrorType) AddDetails(err error, details interface{}) error {
+	if err == nil {
+		return errorType.Details(details)
+	}
+
+	c := new(customError)
+	c.error = errors.WithStack(err)
+	c.userMessage = GetUserMessage(err)
+
+	c.details = append(GetDetails(err), details)
```
Comment (zgalor) [inline errors.go:252]: at the beginning of this func add
```
if details == nil {
return err
}
```

otherwise, you might add `nil` objects to the details slice

File: errors.go [hunk 8]
```diff
@@ -151,6 +152,24 @@ func GetUserMessage(err error) string {
...
+func (c *customError) Details() []interface{} { return c.details }
+
+// GetDetails returns a slice of arbitratry details for all errors
+// If error is not `errorDetailer` returns empty slice
+func GetDetails(err error) []interface{} {
+	if detailedError, ok := err.(errorDetailer); ok {
+		return detailedError.Details()
+	}
+
+	return []interface{}{}
```
Comment (zgalor) [inline errors.go]: returning `nil` is cleaner and has the same impact when you append to it
see https://play.golang.org/p/UVO9hyzVcpn

**Review by zgalor — CHANGES_REQUESTED** [review]

File: errors.go (hunk 5 above)
Comment (zgalor) [reply errors.go]: I think that if someone called AddDetails on a nil error maybe we should return a nil error.
Seems like a misuse to me.
`error()` function on that object would return `""`, `%v` also. I don't know what's the best way to handle that, but I don't think creating empty error with details is a good practice

**Review by tzvatot — COMMENTED** [review]
I disagree @zgalor 
We had this discussion with Wrapf.
Obviously, having Wrapf on a nil err is a misuse. Yet, we got many bugs thrown at us, which caused a lot of noise.
Same goes for AddDetails.
As I see it, Wrapf is just another type of AddDetails, it's just another (more specific) detail - the message.

Consider the following code:
```
err := weberr.Wrapf(e, "some msg")
err = weberr.AddDetails(err, details)
```
If `e` is nil, the above code still works.
If the programmer changes the order of the lines like this:
```
err := weberr.AddDetails(e, details)
err = weberr.Wrapf(err, "some msg")
```
It will throw nil reference.
This basically implies that Wrapf should be used before AddDetails, which make no sense.

Learning from past mistake, I strongly encourage to avoid nil reference due to misuse of the library.

File: errors.go [hunk 9]
```diff
@@ -215,6 +235,36 @@ func (errorType ErrorType) UserErrorf(msg string, args ...interface{}) error {
 	}
 }
 
+// AddDetails adds a details element to an error
+func (errorType ErrorType) AddDetails(err error, details interface{}) error {
+	if err == nil {
+		return errorType.Details(details)
```
Comment (zgalor) [inline errors.go]: I'm having trouble replying to @tzvatot, but I align with what he says. Let's keep it that way

Comment (zgalor) [reply errors.go]: But let's make `Details` private or inline it here.
I don't want to encourage creating details with no error message

File: errors.go (hunk 7 above)
Comment (vkareh) [reply errors.go:252]: Makes sense. Done.

File: errors.go (hunk 8 above)
Comment (vkareh) [reply errors.go]: It works! Didn't realize, but makes sense. Updated it.

File: errors.go (hunk 9 above)
Comment (vkareh) [reply errors.go]: I'm fine with making it private. Done.

File: errors.go [hunk 10]
```diff
@@ -215,6 +235,40 @@ func (errorType ErrorType) UserErrorf(msg string, args ...interface{}) error {
...
+	} else {
+		c.errorType = GetType(err)
+	}
+
+	return c
+}
+
+// details creates a new error with arbitrary details
+func (errorType ErrorType) details(details interface{}) error {
+	return &customError{
```
Comment (zgalor) [inline errors.go:265]: We should add 
``` 
if details == nil {
  return err
}
```

otherwise,  if someone is calling `AddDetails(nil,nil)`
they would get `Details: [{}]` and we don't want that

File: errors_test.go [hunk 11]
```diff
@@ -118,3 +118,50 @@ func TestGetType(t *testing.T) {
...
+	foo := struct{ foo string }{"foo"}
+	bar := struct{ bar string }{"bar"}
+	baz := struct{ baz string }{"baz"}
+	tests := []struct {
+		err      error
+		expected []interface{}
+	}{
+		{nil, []interface{}{}},
+		{io.EOF, []interface{}{}},
+		{AddDetails(nil, foo), []interface{}{foo}},
```
Comment (zgalor) [inline errors_test.go:132]: Please see my previous comment and add a test case `AddDetails(nil, nil)` here

**Review by zgalor — CHANGES_REQUESTED** [review]
Really small change request. Hopefully the last one

File: errors.go (hunk 10 above)
Comment (vkareh) [reply errors.go:265]: good point. adding it now

Comment (vkareh) [reply errors.go:265]: Added it at the begining of `AddDetails`, as `func details()` is only useful when `err == nil`

File: errors_test.go (hunk 11 above)
Comment (vkareh) [reply errors_test.go:132]: Done

**Review by zgalor — APPROVED** [review]
LGTM
